core/chains/evm/gas/universal_estimator_test.go

```diff
@@ -0,0 +1,557 @@
+		client := mocks.NewUniversalEstimatorClient(t)
+		client.On("SuggestGasPrice", mock.Anything).Return(big.NewInt(1), nil).Once()
+		originalGasPrice := assets.NewWeiI(10)
+
+		cfg := gas.UniversalEstimatorConfig{BumpPercent: 50}
+
+		maxPrice := assets.NewWeiI(14)
+		u := gas.NewUniversalEstimator(logger.Test(t), client, cfg, chainID, nil)
+		_, err := u.FetchGasPrice()
+		assert.NoError(t, err)
+		gas, _, err := u.BumpLegacyGas(tests.Context(t), originalGasPrice, gasLimit, maxPrice, nil)
+		assert.NoError(t, err)
+		assert.Equal(t, maxPrice, gas)
+	})
+
+	t.Run("fails if the bumped gas price is lower than the minimum bump percentage", func(t *testing.T) {
+		client := mocks.NewUniversalEstimatorClient(t)
+		client.On("SuggestGasPrice", mock.Anything).Return(big.NewInt(100), nil).Once()
+		originalGasPrice := assets.NewWeiI(100)
+
+		cfg := gas.UniversalEstimatorConfig{BumpPercent: 20}
+
+		// Price will be capped by the max price
+		maxPrice := assets.NewWeiI(101)
+		u := gas.NewUniversalEstimator(logger.Test(t), client, cfg, chainID, nil)
+		_, err := u.FetchGasPrice()
+		assert.NoError(t, err)
+		_, _, err = u.BumpLegacyGas(tests.Context(t), originalGasPrice, gasLimit, maxPrice, nil)
+		assert.Error(t, err)
+	})
+}
+
+func TestUniversalEstimatorGetDynamicFee(t *testing.T) {
+	t.Parallel()
+
+	maxPrice := assets.NewWeiI(100)
+	chainID := big.NewInt(0)
+
+	t.Run("fetches a new dynamic fee when first called", func(t *testing.T) {
+		client := mocks.NewUniversalEstimatorClient(t)
+		baseFee := big.NewInt(5)
+		maxPriorityFeePerGas1 := big.NewInt(33)
+		maxPriorityFeePerGas2 := big.NewInt(20)
+
+		feeHistoryResult := &ethereum.FeeHistory{
+			OldestBlock:  big.NewInt(1),
+			Reward:       [][]*big.Int{{maxPriorityFeePerGas1, big.NewInt(5)}, {maxPriorityFeePerGas2, big.NewInt(5)}}, // first one represents market price and second one connectivity price
+			BaseFee:      []*big.Int{baseFee, baseFee},
+			GasUsedRatio: nil,
+		}
+		client.On("FeeHistory", mock.Anything, mock.Anything, mock.Anything).Return(feeHistoryResult, nil).Once()
+
+		blockHistoryLength := 2
+		cfg := gas.UniversalEstimatorConfig{BlockHistorySize: uint64(blockHistoryLength)}
+		avrgPriorityFee := big.NewInt(0)
+		avrgPriorityFee.Add(maxPriorityFeePerGas1, maxPriorityFeePerGas2).Div(avrgPriorityFee, big.NewInt(int64(blockHistoryLength)))
+		maxFee := (*assets.Wei)(baseFee).AddPercentage(gas.BaseFeeBufferPercentage).Add((*assets.Wei)(avrgPriorityFee))
+
+		u := gas.NewUniversalEstimator(logger.Test(t), client, cfg, chainID, nil)
+		_, err := u.FetchDynamicPrice()
+		assert.NoError(t, err)
+		dynamicFee, err := u.GetDynamicFee(tests.Context(t), maxPrice)
+		assert.NoError(t, err)
+		assert.Equal(t, maxFee, dynamicFee.FeeCap)
+		assert.Equal(t, (*assets.Wei)(avrgPriorityFee), dynamicFee.TipCap)
+	})
+
+	t.Run("fails if BlockHistorySize is zero and tries to fetch new prices", func(t *testing.T) {
```

Comment:
Isn't this test already covered by the one above `fails to start if BlockHistorySize is 0 in EIP-1559` since the estimator would fail to start if eip-1559 is enabled and block history size is 0?

Reply:
Yes, but as you mentioned in the past this is a belt and braces approach, as I really want to avoid the scenario where something gets changed and we have an `index out of range` error in production.